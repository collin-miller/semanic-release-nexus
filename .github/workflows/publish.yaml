name: Release

on:
    push:
        branches:
            - main

jobs:
    release:
        runs-on: ubuntu-latest
        steps:
            - uses: actions/checkout@v2

            # Setup Node.js
            - name: Setup Node.js
              uses: actions/setup-node@v2
              with:
                  node-version: '16'

            - name: Install
              run: yarn

            - name: Build
              run: yarn build

            - name: Release
              env:
                  GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
                  NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
                  NPM_CONFIG_USERCONFIG: .npmrc
              run: npx semantic-release
