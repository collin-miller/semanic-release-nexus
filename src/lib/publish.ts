import AggregateError from 'aggregate-error';
import { Context } from 'semantic-release';
import glob from 'tiny-glob';
import Nexus, { INexusOptions } from './nexus';
import { IPluginConfig, resolveOptions } from './plugin';
import { getFilename } from './get-filename';

export const publish = async (pluginConfig: IPluginConfig, context: Context) => {
    const { logger } = context;
    const options = resolveOptions(pluginConfig, context);
    const nxOptions = {} as INexusOptions;
    if (options.auth.password && options.auth.username) {
        nxOptions.password = options.auth.password;
        nxOptions.username = options.auth.username;
    }
    const nexusClient = new Nexus(options.nexusHost, nxOptions);

    const promises = options.assets.map(async (element) => {
        const [filePath] = await glob(element.path, { filesOnly: true });
        const artifactName = element.name ? element.name : getFilename(filePath);
        logger.log(`Deploying ${filePath} as ${artifactName} to ${options.nexusHost}`);
        return nexusClient.deploy(options.nexusPath, artifactName, filePath);
    });

    await Promise.all(promises).catch((err) => {
        logger.error(err.message);
        throw new AggregateError([err.message]);
    });
};
