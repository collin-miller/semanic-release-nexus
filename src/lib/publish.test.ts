import { describe, expect, it, jest } from '@jest/globals';
import AggregateError from 'aggregate-error';
import glob from 'tiny-glob';
import Nexus from './nexus';
import { publish } from './publish';

jest.mock('./nexus');
jest.mock('tiny-glob');

describe('publish', () => {
    const mockedNexus = Nexus as jest.MockedClass<typeof Nexus>;
    const mockedGlob = glob as jest.Mocked<typeof glob>;

    it('should successfully publish', async () => {
        const mockedDeploy = jest.fn().mockImplementation(() => Promise.resolve(null));
        (mockedNexus as jest.Mock).mockImplementationOnce(() => {
            return {
                deploy: mockedDeploy,
            };
        });
        (mockedGlob as jest.Mock)
            .mockImplementationOnce(() => ['./some-path'])
            .mockImplementationOnce(() => ['./some-path-2']);
        const logMock = jest.fn();
        await expect(
            publish(
                {
                    nexusHost: 'localhost',
                    nexusPath: 'some-repo',
                    assets: [{ path: './some-path' }, { path: './some-path-2', name: 'artifact-2' }],
                },
                { logger: { log: logMock, error: jest.fn() }, env: {} },
            ),
        ).resolves.toBeUndefined();
        expect(logMock).toBeCalledTimes(2);
        expect(logMock.mock.calls).toMatchSnapshot();
        expect(mockedDeploy).toBeCalledTimes(2);
        expect(mockedDeploy).toHaveBeenNthCalledWith(1, 'some-repo', 'some-path', './some-path');
        expect(mockedDeploy).toHaveBeenNthCalledWith(2, 'some-repo', 'artifact-2', './some-path-2');
        expect(mockedNexus).toHaveBeenCalledTimes(1);
        expect(mockedNexus.mock.calls).toMatchSnapshot();
    });

    it('should successfully publish w/ auth', async () => {
        const mockedDeploy = jest.fn().mockImplementation(() => Promise.resolve(null));
        (mockedNexus as jest.Mock).mockImplementationOnce(() => {
            return {
                deploy: mockedDeploy,
            };
        });
        const logMock = jest.fn();
        (mockedGlob as jest.Mock)
            .mockImplementationOnce(() => ['./some-path'])
            .mockImplementationOnce(() => ['./some-path-2']);
        await expect(
            publish(
                {
                    nexusHost: 'localhost',
                    nexusPath: 'some-repo',
                    assets: [{ path: './some-path' }, { path: './some-path-2', name: 'artifact-2' }],
                },
                {
                    logger: { log: logMock, error: jest.fn() },
                    env: { NX_USER: 'some-user', NX_PASSWORD: 'some-password' },
                },
            ),
        ).resolves.toBeUndefined();
        expect(logMock).toBeCalledTimes(2);
        expect(logMock.mock.calls).toMatchSnapshot();
        expect(mockedDeploy).toBeCalledTimes(2);
        expect(mockedDeploy).toHaveBeenNthCalledWith(1, 'some-repo', 'some-path', './some-path');
        expect(mockedDeploy).toHaveBeenNthCalledWith(2, 'some-repo', 'artifact-2', './some-path-2');
        expect(mockedNexus).toHaveBeenCalledTimes(1);
        expect(mockedNexus.mock.calls).toMatchSnapshot();
    });

    it('should raise exception when publish fails', async () => {
        const mockedDeploy = jest
            .fn()
            .mockImplementationOnce(() => Promise.resolve(null))
            .mockImplementationOnce(() => Promise.reject(Error('I hate you')));
        (mockedNexus as jest.Mock).mockImplementationOnce(() => {
            return {
                deploy: mockedDeploy,
            };
        });
        (mockedGlob as jest.Mock)
            .mockImplementationOnce(() => ['./some-path'])
            .mockImplementationOnce(() => ['./some-path-2']);
        const logMock = jest.fn();
        const errorMock = jest.fn();
        await expect(
            publish(
                {
                    nexusHost: 'localhost',
                    nexusPath: 'some-repo',
                    assets: [{ path: './some-path' }, { path: './some-path-2' }],
                },
                { logger: { log: logMock, error: errorMock }, env: {} },
            ),
        ).rejects.toThrowError(AggregateError);
        expect(logMock).toBeCalledTimes(2);
        expect(logMock.mock.calls).toMatchSnapshot();
        expect(errorMock).toBeCalledTimes(1);
        expect(errorMock.mock.calls).toMatchSnapshot();
        expect(mockedDeploy).toBeCalledTimes(2);
        expect(mockedDeploy).toHaveBeenNthCalledWith(1, 'some-repo', 'some-path', './some-path');
        expect(mockedDeploy).toHaveBeenNthCalledWith(2, 'some-repo', 'some-path-2', './some-path-2');
        expect(mockedNexus).toHaveBeenCalledTimes(1);
        expect(mockedNexus.mock.calls).toMatchSnapshot();
    });
});
