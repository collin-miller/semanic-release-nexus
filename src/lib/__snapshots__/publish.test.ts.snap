// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`publish should raise exception when publish fails 1`] = `
Array [
  Array [
    "Deploying ./some-path as some-path to localhost",
  ],
  Array [
    "Deploying ./some-path-2 as some-path-2 to localhost",
  ],
]
`;

exports[`publish should raise exception when publish fails 2`] = `
Array [
  Array [
    "I hate you",
  ],
]
`;

exports[`publish should raise exception when publish fails 3`] = `
Array [
  Array [
    "localhost",
    "some-path",
    "./some-path",
  ],
  Array [
    "localhost",
    "some-path-2",
    "./some-path-2",
  ],
]
`;

exports[`publish should raise exception when publish fails 4`] = `
Array [
  Array [
    "localhost",
    Object {},
  ],
]
`;

exports[`publish should successfully publish 1`] = `
Array [
  Array [
    "Deploying ./some-path as some-path to localhost",
  ],
  Array [
    "Deploying ./some-path-2 as artifact-2 to localhost",
  ],
]
`;

exports[`publish should successfully publish 2`] = `
Array [
  Array [
    "localhost",
    "some-path",
    "./some-path",
  ],
  Array [
    "localhost",
    "artifact-2",
    "./some-path-2",
  ],
]
`;

exports[`publish should successfully publish 3`] = `
Array [
  Array [
    "localhost",
    Object {},
  ],
]
`;

exports[`publish should successfully publish w/ auth 1`] = `
Array [
  Array [
    "Deploying ./some-path as some-path to localhost",
  ],
  Array [
    "Deploying ./some-path-2 as artifact-2 to localhost",
  ],
]
`;

exports[`publish should successfully publish w/ auth 2`] = `
Array [
  Array [
    "localhost",
    "some-path",
    "./some-path",
  ],
  Array [
    "localhost",
    "artifact-2",
    "./some-path-2",
  ],
]
`;

exports[`publish should successfully publish w/ auth 3`] = `
Array [
  Array [
    "localhost",
    Object {
      "password": "some-password",
      "username": "some-user",
    },
  ],
]
`;
